File: services/inventory-service/internal/middleware/correlation.go
package middleware

import (
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

const CorrelationIDHeader = "X-Correlation-ID"

// CorrelationID middleware adds correlation ID to requests
func CorrelationID() gin.HandlerFunc {
	return func(c *gin.Context) {
		correlationID := c.GetHeader(CorrelationIDHeader)
		if correlationID == "" {
			correlationID = uuid.New().String()
		}

		c.Set("correlation_id", correlationID)
		c.Header(CorrelationIDHeader, correlationID)

		c.Next()
	}
}

File: shared/nodejs/middleware/correlation.ts
// Correlation ID middleware for Node.js/Express services
import { Request, Response, NextFunction } from 'express';
import { v4 as uuidv4 } from 'uuid';

export const CORRELATION_ID_HEADER = 'X-Correlation-ID';

declare global {
  namespace Express {
    interface Request {
      correlationId?: string;
    }
  }
}

/**
 * Middleware to extract or generate correlation IDs for request tracing
 * Adds correlation ID to request object and response headers
 */
export function correlationIdMiddleware(req: Request, res: Response, next: NextFunction): void {
  // Extract from header or generate new
  const correlationId = req.headers[CORRELATION_ID_HEADER.toLowerCase()] as string || uuidv4();

  // Attach to request
  req.correlationId = correlationId;

  // Add to response headers
  res.setHeader(CORRELATION_ID_HEADER, correlationId);

  // Add to all logs (if logger supports it)
  if (req.log) {
    req.log = req.log.child({ correlationId });
  }

  next();
}

/**
 * Get correlation ID from request
 */
export function getCorrelationId(req: Request): string {
  return req.correlationId || 'unknown';
}
